pkg/client/testingclient/spy.go

```diff
@@ -0,0 +1,143 @@
+package testingclient
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+)
+
+type SpyCall struct {
+	GVK      schema.GroupVersionKind
+	IsStatus bool
+	Verb     string
+	Key      client.ObjectKey
+	Obj      client.Object
+	List     client.ObjectList
+	Patch    client.Patch
+}
+
+type Spy struct {
+	Delegate client.Client
+	Calls    chan<- SpyCall
+	isStatus bool
+}
+
+var _ client.Client = Spy{}
+
+func (s Spy) Scheme() *runtime.Scheme {
+	return s.Delegate.Scheme()
+}
+
+func (s Spy) RESTMapper() meta.RESTMapper {
+	return s.Delegate.RESTMapper()
+}
+
+func (s Spy) Get(ctx context.Context, key client.ObjectKey, obj client.Object) error {
+	err := s.Delegate.Get(ctx, key, obj)
+	s.Calls <- SpyCall{
+		GVK:  mustGVKForObject(obj, s.Scheme()),
+		Verb: "get",
+		Key:  key,
+		Obj:  obj,
+	}
+	return err
+}
+
+func (s Spy) List(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error {
+	gvk, err := listGVK(list, s.Scheme())
+	if err != nil {
+		return err
+	}
+
+	listErr := s.Delegate.List(ctx, list, opts...)
+	s.Calls <- SpyCall{
+		GVK:  gvk,
+		Verb: "list",
+		List: list,
```

Comment:
You need to DeepCopy everything whenever you store the call, otherwise the object in the saved call might get manipulated later on.

IMHO the functionality here should from an external API POV be part of the reactive client or not exist (internally, it can be a distinct and unexported struct)

Reply:
Thanks, added the `DeepCopyObject()`s. Also had to type-assert the copies back to `client.Object`/`client.ObjectList`. I think that should be safe.